src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs

```diff
@@ -0,0 +1,258 @@
+
+        /// <summary>
+        /// Lock to product <see cref="_documentIdToLastResultId"/> and <see cref="_nextResultId"/>.
+        /// </summary>
+        private readonly object _gate = new object();
+
+        /// <summary>
+        /// Mapping of a document to the last result id we reported for it.
+        /// </summary>
+        private readonly Dictionary<(Workspace workspace, DocumentId documentId), string> _documentIdToLastResultId = new();
+
+        /// <summary>
+        /// The next available id to label results with.
+        /// </summary>
+        private long _nextResultId;
+
+        protected AbstractPullDiagnosticHandler(
+            ILspSolutionProvider solutionProvider,
+            IDiagnosticService diagnosticService)
+        {
+            _solutionProvider = solutionProvider;
+            _diagnosticService = diagnosticService;
+            _diagnosticService.DiagnosticsUpdated += OnDiagnosticsUpdated;
+        }
+
+        public abstract TextDocumentIdentifier? GetTextDocumentIdentifier(TDiagnosticsParams diagnosticsParams);
+
+        /// <summary>
+        /// Gets the progress object to stream results to.
+        /// </summary>
+        protected abstract IProgress<TReport[]>? GetProgress(TDiagnosticsParams diagnosticsParams);
+
+        /// <summary>
+        /// Retrieve the previous results we reported.  Used so we can avoid resending data for unchanged files. Also
+        /// used so we can report which documents were removed and can have all their diagnostics cleared.
+        /// </summary>
+        protected abstract DiagnosticParams[]? GetPreviousResults(TDiagnosticsParams diagnosticsParams);
+
+        /// <summary>
+        /// Returns all the documents that should be processed in the desired order to process them in.
+        /// </summary>
+        protected abstract ImmutableArray<Document> GetOrderedDocuments(RequestContext context);
+        protected abstract TReport CreateReport(TextDocumentIdentifier? identifier, VSDiagnostic[]? diagnostics, string? resultId);
+
+        private void OnDiagnosticsUpdated(object? sender, DiagnosticsUpdatedArgs updateArgs)
+        {
+            if (updateArgs.DocumentId == null)
+                return;
+
+            // Whenever we hear about changes to a document, drop the data we've stored for it.  We'll recompute it as
+            // necessary on the next request.
+            _documentIdToLastResultId.Remove((updateArgs.Workspace, updateArgs.DocumentId));
+        }
+
+        public async Task<TReport[]?> HandleRequestAsync(
+            TDiagnosticsParams diagnosticsParams, RequestContext context, CancellationToken cancellationToken)
+        {
+            // The progress object we will stream reports to.
+            var progress = GetProgress(diagnosticsParams);
+
+            // The buffer we will add results to if our client doesn't support streaming results.
+            using var _ = ArrayBuilder<TReport>.GetInstance(out var reports);
```

Comment:
Is there a known scenario where we need to support both streamed and returned results? Other handlers don't do this, but perhaps we should file issues to add it?

Reply:
Technically per the lsp spec we should be supporting both - the client indicates it by specifying a partial result token
https://microsoft.github.io/language-server-protocol/specification#partialResults

But the VS LSP client supports it

Reply:
Thanks, I will file some issues for the ones I found.

Reply:
Yeah, Etienne said we should support both.  It would be good to ahve an reusable abstraction we could use for this.  I don't actually like what i've done here as it will use both sinks in the streaming case.  It should really only use one or the other, avoiding even making the arraqybuilder if you have an IProgress.